File: backend/src/main/kotlin/com/mongle/service/HomeService.kt
package com.mongle.service

import com.mongle.controller.dto.AvatarGender
import com.mongle.controller.dto.ChipDisplay
import com.mongle.controller.dto.ChipRef
import com.mongle.controller.dto.IntimacyStatus
import com.mongle.controller.dto.MeNode
import com.mongle.controller.dto.PersonNode
import com.mongle.controller.dto.RelationEdge
import com.mongle.controller.dto.RelationMapResponse
import com.mongle.controller.dto.ThrowbackResponse
import com.mongle.domain.ChipType
import com.mongle.domain.Event
import com.mongle.domain.Person
import com.mongle.repository.ChipRepository
import com.mongle.repository.EventRepository
import com.mongle.repository.PersonRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalTime

@Service
@Transactional(readOnly = true)
class HomeService(
    private val personRepository: PersonRepository,
    private val chipRepository: ChipRepository,
    private val personStatsService: PersonStatsService,
    private val eventRepository: EventRepository,
    private val personService: PersonService,
    private val eventService: EventService,
) {
    /**
     * 관계 지도(#40 #41 #42). 내 소유·active 인물을 노드로, 나↔인물을 엣지로 낸다.
     * 노드 정렬은 즐겨찾기 먼저 → 이름순(그래프 표시 결정성). 관계 점수화·서열화가 아니다.
     * 친밀도(#41)는 그 인물 만남 고유 날짜(파생 스탯)로 판정하고, 멀어짐이면 엣지를 흐리게 표시하도록 내린다.
     * 관계태그 필터(#42)는 합집합(OR) — 하나라도 가진 인물을 남긴다(넓혀 보기). 빈 필터는 전체. 흐린 표시는 멀어짐 전용이라 섞지 않고, 조건 밖 인물은 아예 뺀다.
     */
    fun relationMap(userId: Long, filterTagChipIds: List<Long>): RelationMapResponse {
        val today = LocalDate.now()
        val filter = filterTagChipIds.toSet()
        // 관계태그(조인 엔티티)는 필터·표시 양쪽에 쓰이므로 전체 인물분을 한 번에 로드한 뒤 필터한다.
        val all = personRepository.findByOwnerIdAndDeletedAtIsNull(userId)
        val tagChipIdsByPerson = personService.relationTagChipIdsByPerson(all)
        val persons = all
            .filter { filter.isEmpty() || tagChipIdsByPerson[it.id].orEmpty().any(filter::contains) }
            // 최종 타이브레이커 id 오름차순 — 동명이인이 있어도 조회마다 노드 순서가 흔들리지 않게(표시 결정성).
            .sortedWith(compareByDescending<Person> { it.favorite }.thenBy(String.CASE_INSENSITIVE_ORDER) { it.name }.thenBy { it.id })

        val tagDisplays = resolveTagDisplays(tagChipIdsByPerson.values.flatten())
        val nodes = persons.map { person ->
            val stats = personStatsService.statsOf(person)
            val intimacy = IntimacyCalculator.of(stats.meetingDatesDesc, today)
            PersonNode(
                id = requireNotNull(person.id),
                name = person.name,
                profileImageUrl = person.profileImageUrl,
                avatarGender = person.gender?.let { AvatarGender.valueOf(it.name) },
                favorite = person.favorite,
                recordCount = stats.recordCount,
                relationTags = tagChipIdsByPerson[person.id].orEmpty().mapNotNull { id -> tagDisplays[id]?.let { ChipRef(id, it.label, it.color) } },
                intimacy = intimacy,
                firstMetDate = person.firstMetDate,
            )
        }
        val edges = nodes.map { RelationEdge(personId = it.id, distant = it.intimacy.status == IntimacyStatus.DISTANT) }
        return RelationMapResponse(me = MeNode(), nodes = nodes, edges = edges)
    }

    /**
     * 1년 전 오늘 회고(#43). 기록 occurredDate 가 정확히 1년 전(같은 월·일, 연도=올해−1)인 기록만 대상.
     * 카드 문구가 "1년 전 오늘"이라 여러 해 전 기록은 포함하지 않는다.
     * 윤년 2/29: 오늘이 2/29 여도 작년(비윤년)엔 2/29 기록이 있을 수 없어 연도 필터로 자연히 빈 결과가 된다(PRD §5 미노출).
     * 복수면 우선순위 ①즐겨찾기 인물 →②사진 →③기념일 카테고리 →④이른 시각(없으면 뒤) →⑤먼저 남긴 것(id) 로 1건.
     */
    fun throwback(userId: Long): ThrowbackResponse? {
        val today = LocalDate.now()
        val lastYear = today.year - 1
        val candidates = eventRepository.findByOwnerIdAndMonthDay(userId, today.monthValue, today.dayOfMonth)
            .filter { it.occurredDate.year == lastYear }
        if (candidates.isEmpty()) return null

        // 후보 기록의 연결 인물(순서 보존)을 한 번에 로드 — 우선순위 판정(즐겨찾기)·대표 이름에 재사용.
        val personIdsByEvent = eventService.personIdsByEvent(candidates)
        // 우선순위 판정에 필요한 인물(즐겨찾기·대표 이름)을 후보 전체 인물 id 로 한 번에 로드(소프트삭제 포함, 과거 참조 이름 보존).
        val persons = personRepository.findAllById(personIdsByEvent.values.flatten().distinct())
            .mapNotNull { p -> p.id?.let { it to p } }.toMap()
        val favoriteIds = persons.filterValues { it.favorite }.keys
        val anniversaryChipId = chipRepository.findByTypeAndOwnerIdIsNullAndLabelAndDeletedAtIsNull(ChipType.CATEGORY, ANNIVERSARY_CATEGORY_LABEL)?.id

        val selected = candidates.sortedWith(
            compareByDescending<Event> { ev -> personIdsByEvent[ev.id].orEmpty().any(favoriteIds::contains) }
                .thenByDescending { it.photoUrls.isNotEmpty() }
                .thenByDescending { anniversaryChipId != null && it.categoryChipId == anniversaryChipId }
                .thenBy(nullsLast<LocalTime>()) { it.occurredTime }
                .thenBy { it.id },
        ).first()

        val representativeId = personIdsByEvent.getValue(requireNotNull(selected.id)).first()
        return ThrowbackResponse(
            eventId = requireNotNull(selected.id),
            personId = representativeId,
            personName = persons[representativeId]?.name.orEmpty(),
            title = selected.title,
            occurredDate = selected.occurredDate,
            photoUrl = selected.photoUrls.firstOrNull(),
        )
    }

    /**
     * 관계태그 칩 id 목록의 라벨을 한 번에 해석한다 — id 참조라 rename 이 자동 반영되고,
     * 소프트삭제된 칩도 findAllById 로 잡혀 라벨이 유지된다(과거 참조 보존, 00-infra).
     */
    private fun resolveTagDisplays(chipIds: List<Long>): Map<Long, ChipDisplay> = chipRepository.findAllById(chipIds.distinct()).mapNotNull { chip -> chip.id?.let { it to ChipDisplay(chip.label, chip.color) } }.toMap()

    companion object {
        // 회고 우선순위 ③(기념일) 판정용 공통 카테고리 라벨. ChipSeeder 의 CATEGORY 시드와 동일해야 한다.
        private const val ANNIVERSARY_CATEGORY_LABEL = "기념일"
    }
}

File: backend/src/home/home.service.ts
import { Injectable } from '@nestjs/common';
import { ChipService } from '../chips/chip.service';
import { BusinessException } from '../common/exception/business-exception';
import { ErrorCode } from '../common/exception/error-code';
import { bytesToUuid, toNumberId } from '../common/prisma';
import { EventRow, EventService } from '../events/event.service';
import { PersonGender } from '../persons/person.dto';
import { PersonService } from '../persons/person.service';
import { PrismaService } from '../prisma/prisma.service';
import { compareDates, daysBetween, formatDate, formatTime, todayInKst } from '../shared/date';
import { Intimacy, IntimacyStatus, PersonNode, RelationMapResponse, ThrowbackResponse } from './home.dto';

@Injectable()
export class HomeService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly persons: PersonService,
    private readonly events: EventService,
    private readonly chips: ChipService,
  ) {}

  async relationMap(ownerId: Buffer<ArrayBuffer>, filterTagChipIds: bigint[]): Promise<RelationMapResponse> {
    const user = await this.prisma.user.findUnique({ where: { id: ownerId } });
    if (!user) throw new BusinessException(ErrorCode.NOT_FOUND);

    const all = await this.prisma.person.findMany({
      where: { ownerId, deletedAt: null },
    });
    const tagChipIdsByPerson = await this.persons.relationTagChipIdsByPerson(all);
    const filter = new Set(filterTagChipIds.map(String));
    const selected = all
      .filter(
        (person) =>
          filter.size === 0 || (tagChipIdsByPerson.get(person.id) ?? []).some((id) => filter.has(id.toString())),
      )
      .sort((left, right) => {
        if (left.favorite !== right.favorite) return left.favorite ? -1 : 1;
        const byName = compareText(left.name.toLowerCase(), right.name.toLowerCase());
        if (byName !== 0) return byName;
        return left.id < right.id ? -1 : left.id > right.id ? 1 : 0;
      });

    const tagIds = [...new Set([...tagChipIdsByPerson.values()].flat().map((id) => id.toString()))].map(BigInt);
    const tagChips = tagIds.length ? await this.prisma.chip.findMany({ where: { id: { in: tagIds } } }) : [];
    const tagById = new Map(tagChips.map((chip) => [chip.id.toString(), chip]));
    const today = todayInKst();
    const nodes: PersonNode[] = await Promise.all(
      selected.map(async (person) => {
        const stats = await this.persons.statsOf(person);
        return {
          id: toNumberId(person.id),
          name: person.name,
          profileImageUrl: person.profileImageUrl,
          avatarGender: person.gender as PersonGender | null,
          favorite: person.favorite,
          recordCount: stats.recordCount,
          relationTags: (tagChipIdsByPerson.get(person.id) ?? []).flatMap((id) => {
            const chip = tagById.get(id.toString());
            return chip
              ? [
                  {
                    id: toNumberId(id),
                    label: chip.label,
                    color: chip.color,
                  },
                ]
              : [];
          }),
          intimacy: intimacyOf(stats.meetingDatesDesc, today),
          firstMetDate: person.firstMetDate ? formatDate(person.firstMetDate) : null,
        };
      }),
    );
    return {
      me: {
        label: '나',
        id: bytesToUuid(user.id),
        name: user.username,
        profileImageUrl: user.profileImageUrl,
        avatarGender: user.gender as PersonGender | null,
      },
      nodes,
      edges: nodes.map((node) => ({
        personId: node.id,
        distant: node.intimacy.status === IntimacyStatus.DISTANT,
      })),
    };
  }

  async throwback(ownerId: Buffer<ArrayBuffer>): Promise<ThrowbackResponse | null> {
    const today = todayInKst();
    const targetYear = Number(today.slice(0, 4)) - 1;
    const targetMonthDay = today.slice(5);
    const candidates = (
      await this.prisma.event.findMany({
        where: { ownerId, deletedAt: null },
      })
    ).filter((event) => {
      const date = formatDate(event.occurredDate);
      return Number(date.slice(0, 4)) === targetYear && date.slice(5) === targetMonthDay;
    }) as EventRow[];
    if (candidates.length === 0) return null;

    const personIdsByEvent = await this.events.personIdsByEvent(candidates);
    const personIds = [...new Set([...personIdsByEvent.values()].flat().map((id) => id.toString()))].map(BigInt);
    const eventIds = candidates.map((event) => event.id);
    const [people, photos, anniversaryCategoryId] = await Promise.all([
      personIds.length ? this.prisma.person.findMany({ where: { id: { in: personIds } } }) : [],
      this.prisma.eventPhoto.findMany({
        where: { eventId: { in: eventIds } },
        orderBy: [{ eventId: 'asc' }, { photoOrder: 'asc' }],
      }),
      this.chips.anniversaryCategoryId(),
    ]);
    const personById = new Map(people.map((person) => [person.id.toString(), person]));
    const favoriteIds = new Set(people.filter((person) => person.favorite).map((person) => person.id.toString()));
    const photosByEvent = new Map<bigint, string[]>();
    for (const photo of photos) {
      if (photo.url === null) continue;
      const values = photosByEvent.get(photo.eventId) ?? [];
      values.push(photo.url);
      photosByEvent.set(photo.eventId, values);
    }
    const selected = [...candidates].sort((left, right) => {
      const leftFavorite = (personIdsByEvent.get(left.id) ?? []).some((id) => favoriteIds.has(id.toString()));
      const rightFavorite = (personIdsByEvent.get(right.id) ?? []).some((id) => favoriteIds.has(id.toString()));
      if (leftFavorite !== rightFavorite) return leftFavorite ? -1 : 1;
      const leftPhoto = (photosByEvent.get(left.id) ?? []).length > 0;
      const rightPhoto = (photosByEvent.get(right.id) ?? []).length > 0;
      if (leftPhoto !== rightPhoto) return leftPhoto ? -1 : 1;
      const leftAnniversary = anniversaryCategoryId !== null && left.categoryChipId === anniversaryCategoryId;
      const rightAnniversary = anniversaryCategoryId !== null && right.categoryChipId === anniversaryCategoryId;
      if (leftAnniversary !== rightAnniversary) return leftAnniversary ? -1 : 1;
      if (left.occurredTime === null && right.occurredTime !== null) return 1;
      if (left.occurredTime !== null && right.occurredTime === null) return -1;
      if (left.occurredTime !== null && right.occurredTime !== null) {
        const byTime = formatTime(left.occurredTime).localeCompare(formatTime(right.occurredTime));
        if (byTime !== 0) return byTime;
      }
      return left.id < right.id ? -1 : left.id > right.id ? 1 : 0;
    })[0];
    const representativeId = personIdsByEvent.get(selected.id)?.[0];
    if (representativeId === undefined) return null;
    return {
      eventId: toNumberId(selected.id),
      personId: toNumberId(representativeId),
      personName: personById.get(representativeId.toString())?.name ?? '',
      title: selected.title,
      occurredDate: formatDate(selected.occurredDate),
      photoUrl: photosByEvent.get(selected.id)?.[0] ?? null,
    };
  }
}

function intimacyOf(meetingDatesDesc: string[], today: string): Intimacy {
  const lastMet = meetingDatesDesc[0] ?? null;
  const daysSinceLastMeet = lastMet === null ? null : daysBetween(lastMet, today);
  if (meetingDatesDesc.length < 2) {
    return {
      status: IntimacyStatus.UNKNOWN,
      averageIntervalDays: null,
      daysSinceLastMeet,
    };
  }
  const ascending = [...meetingDatesDesc].sort(compareDates);
  const intervals = ascending.slice(1).map((date, index) => daysBetween(ascending[index], date));
  const average = intervals.reduce((sum, days) => sum + days, 0) / intervals.length;
  return {
    status: daysSinceLastMeet! > average * 2 ? IntimacyStatus.DISTANT : IntimacyStatus.NORMAL,
    averageIntervalDays: Math.round(average),
    daysSinceLastMeet,
  };
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}
